src/main/java/slimeknights/tconstruct/library/materials/Material.java

```diff
@@ -0,0 +1,201 @@
+    // lowercases and removes whitespaces
+    this.identifier = new MaterialId(identifier.getNamespace(), Util.sanitizeLocalizationString(identifier.getPath()));
+    this.fluid = fluid;
+    this.craftable = craftable;
+    this.shardItem = shardItem;
+  }
+
+  @Override
+  public MaterialId getIdentifier() {
+    return identifier;
+  }
+
+  @Override
+  public boolean isCraftable() {
+    return this.craftable;
+  }
+
+  @Override
+  public Fluid getFluid() {
+    return fluid;
+  }
+
+  /* Stats */
+
+  /**
+   * Do not use this function directly stats. Use TinkerRegistry.addMaterialStats instead.
+   */
+  public Material addStats(IMaterialStats materialStats) {
+    this.stats.put(materialStats.getIdentifier(), materialStats);
+    return this;
+  }
+
+  /**
+   * Returns the material stats of the given type of this material.
+   *
+   * @param partType Identifier of the material.
+   * @param <T>      Type of the Stats are determined by return value. Use the correct
+   * @return The stats found or null if none present.
+   */
+  @Override
+  @SuppressWarnings("unchecked")
+  public <T extends IMaterialStats> Optional<T> getStatsForType(MaterialStatType partType) {
+    return Optional.ofNullable((T) stats.get(partType));
+  }
+
+  @Override
+  public Collection<IMaterialStats> getAllStats() {
+    return stats.values();
+  }
+
+  /* Traits */
+
+  /**
+   * Adds the trait as the default trait, will be used if no more specific one is present time.
+   */
+  public Material addTrait(ITrait materialTrait) {
+    return addTrait(materialTrait, null);
+  }
+
+  /**
+   * Adds the trait to be added if the specified stats are used.
+   */
+  public Material addTrait(ITrait materialTrait, ResourceLocation forStatsType) {
+    // todo: we don't register traits automatically on addition anymore, check if all are present
+    getStatTraits(forStatsType).add(materialTrait);
+    return this;
+  }
+
+  @Override
+  public List<ITrait> getAllTraitsForStats(MaterialStatType partType) {
+    List<ITrait> traits = traitsByStats.get(partType);
+
+    if (traits == null) {
+      return traitsByStats.getOrDefault(null, ImmutableList.of());
+    } else {
+      return traits;
+    }
+  }
+
+  /**
+   * Obtains the list of traits for the given stat, creates it if it doesn't exist yet.
+   */
+  @SuppressWarnings("WeakerAccess")
+  protected List<ITrait> getStatTraits(ResourceLocation forStatsType) {
+    if (!this.traitsByStats.containsKey(forStatsType)) {
+      // linked list since we're only ever iterating over the list
+      this.traitsByStats.put(forStatsType, new LinkedList<>());
+    }
+    return this.traitsByStats.get(forStatsType);
+  }
+
+  @Override
+  public Collection<ITrait> getAllTraits() {
+    return traitsByStats.values().stream()
+      .flatMap(Collection::stream)
+      .collect(Collectors.toList());
+  }
+
+  public void setShard(ItemStack stack) {
```

Comment:
Do we still need shards? I thought we don't have any parts that cost 0.5 material anymore.

Reply:
Good point, I need to check where else it's used. It'd actually be great if we got rid of the "leftovers" part in tool building.

Reply:
One case that is worth cosidering is using a >1 material piece (like a log) to craft a smaller part.
* We could just require every material have a part of <=1 and use that part as the "shard", but then we still have to deal with leftovers which we want to avoid.
* Another option would be requiring an exact match, e.g. logs can only be used if you want a part costing 4
* We could discard excess material. This would be the simpliest to code, but would cause questions of why a log gives only 1 tool rod.
* Finally, we could return multiple of the part. Math gets a bit weird if not an exact match, so maybe discard excess beyond a single use of the material? Overall probably the toughest to code.

Personally, I like giving multiple matches then discarding excess, but that would be complicated.

Reply:
Last option definitively in this case. Use what you can and discard the rest. In most cases you'll be using "1 worth" items anyway. And you can always craft logs/ironblocks etc. into their smaller parts.